packages/ckeditor5-widget/src/utils.ts

```diff
@@ -453,20 +453,36 @@ function addSelectionHandle( widgetElement: ViewContainerElement, writer: Downca
  * Starting from a DOM resize host element (an element that receives dimensions as a result of resizing),
  * this helper returns the width of the found ancestor element.
  *
- * **Note**: This helper searches up to 5 levels of ancestors only.
+ *	* It searches up to 5 levels of ancestors only.
+ *	* It will skip search if passed element (or it's parent) has `contenteditable` attribute or is root document editable.
+ *	  It's crucial for features such like pagination because editable root element might have different width than whole editor itself.
  *
  * @param domResizeHost Resize host DOM element that receives dimensions as a result of resizing.
  * @returns Width of ancestor element in pixels or 0 if no ancestor with a computed width has been found.
  */
 export function calculateResizeHostAncestorWidth( domResizeHost: HTMLElement ): number {
+	const getElementComputedWidth = ( element: HTMLElement ) => {
+		const { width, paddingLeft, paddingRight } = element.ownerDocument.defaultView!.getComputedStyle( element! );
+
+		return parseFloat( width ) - ( parseFloat( paddingLeft ) || 0 ) - ( parseFloat( paddingRight ) || 0 );
+	};
+
+	if ( isMaximumResizeHostAncestorElement( domResizeHost ) ) {
+		return getElementComputedWidth( domResizeHost );
+	}
+
 	const domResizeHostParent = domResizeHost.parentElement;
 
 	if ( !domResizeHostParent ) {
 		return 0;
 	}
 
 	// Need to use computed style as it properly excludes parent's paddings from the returned value.
-	let parentWidth = parseFloat( domResizeHostParent!.ownerDocument.defaultView!.getComputedStyle( domResizeHostParent! ).width );
+	let parentWidth = getElementComputedWidth( domResizeHostParent! );
+
+	if ( isMaximumResizeHostAncestorElement( domResizeHostParent ) ) {
```

Comment:
This is only masking the actual root of the issue in `ImageResizeHandles`. Check out the current implementation at https://github.com/ckeditor/ckeditor5/blob/67d2d60762ca1334beda744ce8f0fd28812ce1f1/packages/ckeditor5-image/src/imageresize/imageresizehandles.ts#L103 and see what the output is for various combinations of editors (decoupled, classic) and images (block, inline). 

Long story short, it returns paragraphs for inline images and editing roots for block images (that exist directly in a root). **This is wrong**. According to `_getResizeHost()` in `Resizer` class

> Resize host is an object that receives dimensions which are the result of resizing.

https://github.com/ckeditor/ckeditor5/blob/67d2d60762ca1334beda744ce8f0fd28812ce1f1/packages/ckeditor5-widget/src/widgetresize/resizer.ts#L413-L422

An editing root does not receive dimensions. `<figure>` does (block image) or `<span>` (inline image).